modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java

```diff
@@ -140,6 +149,56 @@ public void testAddStructureWithDuplicateKey() throws Exception {
 			DDMStructureConstants.TYPE_DEFAULT);
 	}
 
+	@Test
+	public void testAddStructureWithExternalReferenceCode() throws Exception {
+		String externalReferenceCode = RandomTestUtil.randomString();
+
+		DDMStructure ddmStructure1 = _addStructure(
+			externalReferenceCode,
+			PortalUtil.getClassNameId(DDL_RECORD_SET_CLASS_NAME),
+			group.getGroupId(), RandomTestUtil.randomString(),
+			TestPropsValues.getUserId());
+
+		Assert.assertEquals(
+			externalReferenceCode, ddmStructure1.getExternalReferenceCode());
+
+		AssertUtils.assertFailure(
+			DuplicateDDMStructureExternalReferenceCodeException.class,
+			StringBundler.concat(
+				"Duplicate structure with the external reference code \"",
+				externalReferenceCode, "\" and class name ID \"",
+				ddmStructure1.getClassNameId(), "\" on site ID \"",
+				ddmStructure1.getGroupId(), "\""),
+			() -> _addStructure(
+				externalReferenceCode, ddmStructure1.getClassNameId(),
+				ddmStructure1.getGroupId(), RandomTestUtil.randomString(),
+				TestPropsValues.getUserId()));
+
+		// With same external reference code and different class name ID
```

Comment:
// Same external reference code but different class name ID

Reply:
Good to know, I already did this change!